internal/provider/provider.go

```diff
@@ -47,15 +42,58 @@ type cacheKey struct {
 	method     string
 }
 
-func (p *provider) login(ctx context.Context, client *api.Client, vaultKubernetesMountPath, roleName, jwt string) (string, error) {
-	p.logger.Debug("vault: performing vault login...")
+func (p *provider) createJWTToken(ctx context.Context, podInfo config.PodInfo) (string, error) {
+	p.logger.Debug("creating service account token bound to pod",
+		"namespace", podInfo.Namespace,
+		"serviceAccountName", podInfo.ServiceAccountName,
+		"podName", podInfo.Name,
+		"podUID", podInfo.UID)
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return "", err
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return "", err
+	}
+
+	ttl := int64((15 * time.Minute).Seconds())
+	resp, err := clientset.CoreV1().ServiceAccounts(podInfo.Namespace).CreateToken(ctx, podInfo.ServiceAccountName, &authenticationv1.TokenRequest{
+		Spec: authenticationv1.TokenRequestSpec{
```

Comment:
Wondering if we should add labels to reflect these were created by the CSI provider? It might be helpful if you want to query who owns them.

Reply:
It is assumed that the token JWT created contains the requested Pod's namespace and service account. If so, it should be included in the Kubernetes Auth method metadata, so I think we should also include testing with the Templated policy. This proves that namespace-scoped access control is possible using the Templated policy.

Reply:
On the other hand, even if you prepare a ClusterRole and request a token from the API server, `MountRequest` contains the ServiceAccount Token specified by `nodePublishSecretRef`. ([source](https://github.com/kubernetes-sigs/secrets-store-csi-driver/blob/master/provider/v1alpha1/service.proto#L49)) If you specify a secret that the pod itself can refer to in the inline volume of the pod, it can be done without requesting a token to API server. you do not need to add a new ClusterRole. What do you think?

Reply:
> Wondering if we should add labels to reflect these were created by the CSI provider?

Is this on the assumption that a k8s secret is created for the token? I just had a check, and by default it doesn't get mounted into a secret, it just lives in the API store until it expires 15m later.

> It is assumed that the token JWT created contains the requested Pod's namespace and service account. If so, it should be included in the Kubernetes Auth method metadata, so I think we should also include testing with the Templated policy. This proves that namespace-scoped access control is possible using the Templated policy.

Sorry I'm not sure if I've understood this comment correctly, but some hopefully relevant comments:
* We do explicitly bind a vault policy to a specific [service account + namespace] in the tests, but vault doesn't care which service account token you use, as long as the issuer you put in the vault k8s auth endpoint matches what it finds in the JWT
* There isn't a negative test for a service account from the wrong namespace, but I guess IMO that would feel more like testing kubernetes/vault k8s auth method than CSI.

> On the other hand, even if you prepare a ClusterRole and request a token from the API server, MountRequest contains the ServiceAccount Token specified by nodePublishSecretRef. (source) If you specify a secret that the pod itself can refer to in the inline volume of the pod, it can be done without requesting a token to API server. you do not need to add a new ClusterRole. What do you think?

Thanks for this comment, it's a very interesting idea, and not one I had considered in depth before. Some thoughts on the tradeoff of this against the PR's current approach:
* `+` No ClusterRole required, fewer privileges for the provider
* `-` Harder to deploy as you don't know a service account's k8s secret name until the SA is deployed
* `-` Assuming you use the default service account token created by kubernetes, then the JWT has no expiry
* `+/-` JWT issuer remains as the default `kubernetes/serviceaccount`, it's one less thing for vault operators to configure but it also makes the validation check less meaningful

Ultimately, I would want to replace either solution with the ability for the CSI driver to request a bound service account token as part of the `NodePublishVolume` request: [KEP link](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/1855-csi-driver-service-account-token/README.md). That's behind a feature gate in alpha for 1.20, so it's very new, but it would get us the best of both worlds.

Given the tradeoffs, I'd suggest for now we open an issue that tracks utilising `nodePublishSecretRef` as a separate feature, and see how quickly the upstream improvement comes down the pipe in the meantime. What do you think?

Reply:
> There isn't a negative test for a service account from the wrong namespace, but I guess IMO that would feel more like testing kubernetes/vault k8s auth method than CSI.

Yes, but the service account toke is created by the vault provider's request so that we should confirm that we can use entity metadata in the policy. btw, I was talking about [policy templating](https://learn.hashicorp.com/tutorials/vault/policy-templating) with vault provider + Kubernetes auth method. 

> Ultimately, I would want to replace either solution with the ability for the CSI driver to request a bound service account token as part of the NodePublishVolume request: KEP link. That's behind a feature gate in alpha for 1.20, so it's very new, but it would get us the best of both worlds.

Yes, I also follow the KEP that will be implemented secrets-store-csi-driver, not the provider itself  https://github.com/kubernetes-sigs/secrets-store-csi-driver/issues/453 :) I assume this feature is an alternative way to authenticate with requesting pod for Kubernetes that is older than 1.20(and for the probable majority of users). so that I thought it was not good to give an extra role to the provider side, even though alternative way.

> - Harder to deploy as you don't know a service account's k8s secret name until the SA is deployed
> - Assuming you use the default service account token created by kubernetes, then the JWT has no expiry
> +/- JWT issuer remains as the default kubernetes/serviceaccount, it's one less thing for vault operators to configure but it also makes the validation check less meaningful

I agree with that. but we can manage not to use the default service account by [`bound_service_account_names`](https://www.vaultproject.io/api-docs/auth/kubernetes#bound_service_account_names). I'm not sure yet HashiCorp's recommended configuration for Kubernetes auth method. Sure, the JWT doesn't have an expiration date, but the vault token also has a TTL, so I didn't think it would be a big deal. I'm still not sure what kind of case the JWT needs to expire.

> Given the tradeoffs, I'd suggest for now we open an issue that tracks utilising nodePublishSecretRef as a separate feature, and see how quickly the upstream improvement comes down the pipe in the meantime. What do you think?

Sure. I want to be clear better way for the requested pod's service account. and will see the activity on csi-driver.

Reply:
> Yes, but the service account toke is created by the vault provider's request so that we should confirm that we can use entity metadata in the policy.

Oh I see, yeah that seems like a nice test. I'll add that.

> Assuming you use the default service account token created by kubernetes

On this point, I meant the default token of a non-default service account, not the `default` service account. e.g. if you create a service account called `nginx-sa`, kubernetes will create a k8s secret called something like `nginx-sa-token-xyzxy`, where the suffix isn't known until the service account is deployed. My assumption was that this is the k8s secret you would reference with `nodeSecretPublishRef`. So operators would have to split their deployment into two parts. One to deploy the service account and query the created token's name, and then another to deploy the pod/volume spec that references that token name.

> I'm still not sure what kind of case the JWT needs to expire.

This is just a consideration for defence in depth. When breaches happen, secrets that live forever can make them higher impact and harder to respond to, so the less we depend on them the better.

Reply:
We can [manually create API token using existing ServiceAccount with the fixed name](https://kubernetes.io/docs/reference/access-authn-authz/service-accounts-admin/#to-create-additional-api-tokens) so that operator doesn't need to split their deployment into two parts.

I agree that the service account token will live forever unless the user deletes it. Nevertheless, using ClusterRole with the provider feels like a bad practice. I would agree if the CSI driver will provide a short-term service account token through the `MountRequest` call. However, it is not certain if secrets-store-csi-driver needs that feature. it seems to be a good feature for users prior to v1.20. but this is a rough idea. I'm still thinking to support reading `nodePublishSecretRef` rather than cluster-wide permission as an option.